crates/runtime/src/externref.rs

```diff
@@ -393,6 +461,704 @@ impl Deref for VMExternRef {
+    /// current bump chunk is full.
+    ///
+    /// This method is infallible, and will allocate an additional bump chunk if
+    /// necessary.
+    #[inline(never)]
+    pub fn insert_slow_path(&self, externref: VMExternRef) {
+        let externref = match self.try_insert(externref) {
+            Ok(()) => return,
+            Err(x) => x,
+        };
+
+        {
+            let mut chunks = self.chunks.borrow_mut();
+
+            let new_size = chunks.last().unwrap().len() * 2;
+            let new_chunk = Self::new_chunk(new_size);
+
+            unsafe {
+                let next = new_chunk.as_ptr() as *mut TableElem;
+                debug_assert!(!next.is_null());
+                *self.next.get() = NonNull::new_unchecked(next);
+
+                let end = next.add(new_chunk.len());
+                debug_assert!(!end.is_null());
+                self.end.set(NonNull::new_unchecked(end));
+            }
+
+            chunks.push(new_chunk);
+        }
+
+        self.try_insert(externref)
+            .expect("insertion should always succeed after we allocate a new chunk");
+    }
+
+    fn num_filled_in_last_chunk(&self, chunks: &[Box<[TableElem]>]) -> usize {
+        let last_chunk = chunks.last().unwrap();
+        let next = unsafe { *self.next.get() };
+        let end = self.end.get();
+        let num_unused_in_last_chunk =
+            ((end.as_ptr() as usize) - (next.as_ptr() as usize)) / mem::size_of::<usize>();
+        last_chunk.len().saturating_sub(num_unused_in_last_chunk)
+    }
+
+    fn elements(&self, mut f: impl FnMut(&VMExternRef)) {
+        // Every chunk except the last one is full, so we can simply iterate
+        // over all of their elements.
+        let chunks = self.chunks.borrow();
+        for chunk in chunks.iter().take(chunks.len() - 1) {
+            for elem in chunk.iter() {
+                if let Some(elem) = unsafe { &*elem.get() } {
+                    f(elem);
+                }
+            }
+        }
+
+        // The last chunk is not all the way full, so we only iterate over its
+        // full parts.
+        let num_filled_in_last_chunk = self.num_filled_in_last_chunk(&chunks);
+        for elem in chunks.last().unwrap().iter().take(num_filled_in_last_chunk) {
+            if let Some(elem) = unsafe { &*elem.get() } {
+                f(elem);
+            }
+        }
+    }
+
+    fn insert_precise_stack_root(&self, root: NonNull<VMExternData>) {
+        let mut precise_stack_roots = self.precise_stack_roots.borrow_mut();
+        if precise_stack_roots.insert(root) {
+            // If this root was not already in the set, then we need to
+            // increment its reference count, so that it doesn't get freed in
+            // `reset` when we're overwriting old bump allocation table entries
+            // with new ones.
+            unsafe {
+                root.as_ref().increment_ref_count();
```

Comment:
Because of this, could the activation table have a `Drop` which asserts that the `precise_stack_roots` set is empty? Otherwise we'd leak data accidentally b/c there's no dtor which drains the map.

(or maybe we should also assert that all chunks are `None` since they don't get frobbed on `Drop` either?)